src/main/java/com/aodocs/partialresponse/fieldsexpression/FieldsExpression.java

```diff
@@ -19,55 +19,71 @@
  */
 package com.aodocs.partialresponse.fieldsexpression;
 
+import com.google.common.collect.ImmutableList;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
-import com.google.common.collect.ImmutableList;
-
 /**
- * Describes a fields expression for partial response, as described there: https://developers.google.com/drive/api/v3/performance#partial-response
+ * Describes a fields expression for partial response, as described there: 
+ * https://developers.google.com/drive/api/v3/performance#partial-response
  */
 public final class FieldsExpression {
 	
 	public static FieldsExpression parse(String fieldsExpression) {
 		return new FieldsExpression(fieldsExpression);
 	}
 	
-	private final String fieldExpression;
-	private final List<ImmutableList<String>> fieldsExpressionPaths;
+	private final String fieldsExpression;
+	private final List<ImmutableList<String>> allPaths;
+	private final List<ImmutableList<String>> collapsedPaths;
+	private final FieldsExpressionTree tree;
 	
 	private FieldsExpression(String fieldsExpression) {
-		this.fieldExpression = fieldsExpression;
-		this.fieldsExpressionPaths = Parser.parse(fieldsExpression);
+		this.fieldsExpression = fieldsExpression;
+		this.allPaths = Parser.parse(fieldsExpression);
+		FieldsExpressionNode.Builder builder = FieldsExpressionNode.Builder.createRoot();
+		this.collapsedPaths = collapsePaths(allPaths)
+				.peek(builder::getOrAddBranch).collect(Collectors.toList());
+		this.tree = new FieldsExpressionTree(builder.getNode());
 	}
-	
+
+	public List<ImmutableList<String>> getFieldPaths() {
+		return collapsedPaths;
+	}
+
+	public FieldsExpressionTree getFilterTree() {
+		return tree;
+	}
+
 	public boolean isValidAgainst(FieldsExpressionTree schema) {
-		return asPathNodes().allMatch(root -> schema.contains(new FieldsExpressionTree(root)));
+		//we check validity with all paths, as collapsePaths might have removed invalid paths
+		return allPaths.stream()
+				.map(path -> createTreeFromPath(path))
+				.allMatch(root -> schema.contains(root));
 	}
-	
-	private Stream<FieldsExpressionNode> asPathNodes() {
-		return fieldsExpressionPaths.stream().map(path -> {
-			FieldsExpressionNode.Builder builder = FieldsExpressionNode.Builder.createRoot();
-			builder.getOrAddBranch(path);
-			return builder.getNode();
-		});
+
+	public boolean contains(FieldsExpressionTree testedTree) {
```

Comment:
We don't really check containment here, as we use contains() both ways ... 
"overlapsWith" might be a better name.

Reply:
I agree